### PR TITLE
Jam fix

### LINKED ISSUE #2891: Add Jam to Raspiblitz WebUI
For v1.7.2 the JoinMarket WebUI was merged but just available thru commandline install - see #2747 

For v1.8 update and integrate into SSH and Web one-click install menus.

### PR DESCRIPTION
**Pull request by openoms**
fixes #2891

- [x] add Jam to menu and recovery
- [x] expose the order-book on /obwatch with nginx. needs: https://github.com/openoms/joininbox/pull/98
- [x] make the display (00infoBlitz.sh) recognise when the YG is running with Jam
- [x] use dergigi-s key to verify
- [x] add to update menu
- [x] test

### CODE REVIEW

**Comment by openoms** [comment]
FYI @dergigi 

Should we rename the script as well to 
`bonus.joinmarket-jam.sh`
or
just `bonus.jam.sh` if the branding will be strong ?

**Comment by dergigi** [comment]
Awesome! Thank you for looking into fixing this.

I'd vote for `bonus.jam.sh` in terms of name. Nice and simple 👍

File: home.admin/config.scripts/bonus.jam.sh [hunk 1]
```diff
@@ -33,7 +33,7 @@ if [ "$1" = "menu" ]; then
     if [ "${runBehindTor}" = "on" ] && [ ${#toraddress} -gt 0 ]; then
       # Info with Tor
       sudo /home/admin/config.scripts/blitz.display.sh qr "${toraddress}"
-      whiptail --title " JoinMarket Web UI " --msgbox "Open in your local web browser:
+      whiptail --title " JAM (JoinMarket Web UI) " --msgbox "Open in your local web browser:
```
Comment (dergigi) [inline home.admin/config.scripts/bonus.jam.sh]:
```suggestion
      whiptail --title " Jam (JoinMarket Web UI) " --msgbox "Open in your local web browser:
```

File: home.admin/config.scripts/bonus.jam.sh [hunk 2]
```diff
@@ -42,7 +42,7 @@ Hidden Service address for Tor Browser (see LCD for QR):\n${toraddress}
       sudo /home/admin/config.scripts/blitz.display.sh hide
     else
       # Info without Tor
-      whiptail --title " JoinMarket Web UI " --msgbox "Open in your local web browser & accept self-signed cert:
+      whiptail --title " JAM (JoinMarket Web UI) " --msgbox "Open in your local web browser & accept self-signed cert:
```
Comment (dergigi) [inline home.admin/config.scripts/bonus.jam.sh]:
```suggestion
      whiptail --title " Jam (JoinMarket Web UI) " --msgbox "Open in your local web browser & accept self-signed cert:
```

**Review by dergigi — APPROVED** [review]
NIT: JAM -> Jam

Looks good to me, utACK ✅

**Comment by rootzoll** [comment]
@openoms ready to merge?

**Comment by openoms** [comment]
> @openoms ready to merge?

yes, the Jam install and update is working now.
(will have a new Joininbox  release with the latest JM to complete this in an unrelated change)

**Comment by openoms** [comment]
ok, all Jam, JoinMarket and Joininbox are on the latest

**Comment by theborakompanioni** [comment]
> > @openoms ready to merge?
> 
> yes, the Jam install and update is working now.

Just a quick question @openoms: Shouldn't the websocket location be `/jmws` and proxied to `28283` (instead of `/ws` to `28183`)? e.g. [this is how it's done in the nginx config](https://github.com/joinmarket-webui/jam-docker/blob/v0.1.0-clientserver-v0.9.8/standalone/nginx/default.conf#L69-L83) of the jam-standalone docker image (<- any and all suggestions welcome on this approach).

**Comment by openoms** [comment]
> > > @openoms ready to merge?
> > 
> > 
> > yes, the Jam install and update is working now.
> 
> Just a quick question @openoms: Shouldn't the websocket location be `/jmws` and proxied to `28283` (instead of `/ws` to `28183`)? e.g. [this is how it's done in the nginx config](https://github.com/joinmarket-webui/jam-docker/blob/v0.1.0-clientserver-v0.9.8/standalone/nginx/default.conf#L69-L83) of the jam-standalone docker image (<- any and all suggestions welcome on this approach).

Thank you for reviewing, not sure tbh, but will test. It was originally added by @dennisreimann in https://github.com/rootzoll/raspiblitz/pull/2747. Did the config change since?

**Comment by theborakompanioni** [comment]
> Thank you for reviewing, not sure tbh, but will test. It was originally added by @dennisreimann in #2747. Did the config change since?

Yes, it changed quite a bit since then (guess this was v0.0.2). But afaik, websockets have always been served on port 28283.
It seems with this config, it will not support the orderbook or the logs feature to be available in Jam on RaspiBlitz (which is generally "okay", Jam will display an error message in the orderbook view, and not display the logs feature at all..).
I guess enabling the log feature is rather difficult, but enabling the orderbook would be really nice. Pardon the question, but does Joininbox run the `ob-watch` service?

**Comment by openoms** [comment]
corrected the websockets in [2c2f3c7](https://github.com/rootzoll/raspiblitz/pull/3321/commits/2c2f3c73dcf520a868c13130cea5ab3663e3ddd0) @theborakompanioni 

>does Joininbox run the ob-watch service?

not by default, only if selected from the menu. It starts a service called: `ob-watcher.service` on the localhost port 62601

Tor address: `sudo cat "$HiddenServiceDir"/ob-watcher/hostname`

where:
```
checkDirEntry=$(grep -c "HiddenServiceDir" < /home/joinmarket/joinin.conf)
if [ "$checkDirEntry" -eq 0 ]; then
  if [ -d "/mnt/hdd/tor" ] ; then
    HiddenServiceDir="/mnt/hdd/tor"
  else
    HiddenServiceDir="/var/lib/tor"
  fi
  echo "HiddenServiceDir=$HiddenServiceDir" >> /home/joinmarket/joinin.conf
fi
```

There would be no problem with running the ob-watcher always. Should we switch it on for Jam so it is reachable on 62601?

**Comment by theborakompanioni** [comment]
> corrected the websockets in [2c2f3c7](https://github.com/rootzoll/raspiblitz/pull/3321/commits/2c2f3c73dcf520a868c13130cea5ab3663e3ddd0) @theborakompanioni

Nice.

Did you try it, though? Has it been successful? I am often unsure whether it works as expected just by looking at configs, always need to verify.
In Jam's docker the config says `location = /jmws` as well as trailing slash at the end of the `proxy_pass` directive: `proxy_pass https://jmwalletd_ws_backend/;` (guess that is not important as JM will except any path on the websocket port).

> > does Joininbox run the ob-watch service?
> 
> not by default, only if selected from the menu. It starts a service called: `ob-watcher.service` on the localhost port 62601
> 
> Tor address: `sudo cat "$HiddenServiceDir"/ob-watcher/hostname`
> 
> where:
> 
> ```
> checkDirEntry=$(grep -c "HiddenServiceDir" < /home/joinmarket/joinin.conf)
> if [ "$checkDirEntry" -eq 0 ]; then
>   if [ -d "/mnt/hdd/tor" ] ; then
>     HiddenServiceDir="/mnt/hdd/tor"
>   else
>     HiddenServiceDir="/var/lib/tor"
>   fi
>   echo "HiddenServiceDir=$HiddenServiceDir" >> /home/joinmarket/joinin.conf
> fi
> ```
> 
> There would be no problem with running the ob-watcher always. Should we switch it on for Jam so it is reachable on 62601?

Hmm.. whatever you want! I don't think the default needs to change necessarily.. but it would be nice if it worked when a users opts-in. So proxying the request would be nice.. Jam expects the api to be available on path [`/obwatch/`](https://github.com/joinmarket-webui/jam-docker/blob/v0.1.0-clientserver-v0.9.8/standalone/nginx/default.conf#L85-L98).

**Comment by openoms** [comment]
the order book should be available now in Jam when using the latest joininbox scripts.

Are there more nginx config changes to be added for the [logs](https://github.com/joinmarket-webui/jam-docker/blob/v0.1.0-clientserver-v0.9.8/standalone/nginx/default.conf#L129) @theborakompanioni ?

**Comment by theborakompanioni** [comment]
> the order book shouls be available now in Jam when using the latest joininbox scripts.

Awesome! :rocket: :rocket: :rocket: 

> Are there more nginx config changes to be added for the [logs](https://github.com/joinmarket-webui/jam-docker/blob/v0.1.0-clientserver-v0.9.8/standalone/nginx/default.conf#L129) @theborakompanioni ?

Nope, I guess that would be too much of a hassle, log files are named differently in the docker image, hence it would need more customization. Would you be okay with addressing this some other time in the future? Will think about how to generalize this. In the meantime, I guess RaspiBlitz users will be capable of viewing the log files for themselves. 

(This feature has only been thought of as a temporary solution for beginners on Umbrel/Citadel to provide better/more info in bug reports.)

Thanks @openoms :orange_heart:

**Comment by openoms** [comment]
Great!
Yes I agree, the logs can be easily checked in the CLI and exposed in the Joininbox menu also.

Will just wait to finalize a next minor Joininbox release to include the small change here in case something else comes up.

**Comment by openoms** [comment]
almost there, but this should be automated (can be in the `precheck` part to run when the joinmarket-api is (re)started):
image
did we have any consensus what to use here?

From https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/jmclient/jmclient/configure.py#L94:
```
# Maximum absolute coinjoin fee in satoshi to pay to a single
# market maker for a transaction. Both the limits given in
# max_cj_fee_abs and max_cj_fee_rel must be exceeded in order
# to not consider a certain offer.
#max_cj_fee_abs = x
# Maximum relative coinjoin fee, in fractions of the coinjoin value
# e.g. if your coinjoin amount is 2 btc (200 million satoshi) and
# max_cj_fee_rel = 0.001 (0.1%), the maximum fee allowed would
# be 0.002 btc (200 thousand satoshi)
#max_cj_fee_rel = x
```

default for Makers in Jam:
image
this comes to 30k sats for a Maker when sending 1BTC.
Since the values are randomized and both needs to be exceeded for the offer to be not considered how about:
```
# max_cj_fee_abs between 5000 - 10000 sats
max_cj_fee_abs = $(shuf -i 5000-10000 -n1)

# max_cj_fee_rel between 0.01 - 0.03%
max_cj_fee_rel = 0.000$((RANDOM%3+1))
```
for the max amount and relative fee considered for a Maker?

EDIT: settled on the randomized values as discusssed on TG: https://t.me/JoinMarketWebUI/4554

**Comment by theborakompanioni** [comment]
@openoms Release `v0.1.1` should drop in a few days. To wait, or not to wait?

**Comment by openoms** [comment]
Not in a rush with this, can wait a few days for sure. Can also update afterwards up to the v1.8.1 release.

**Comment by theborakompanioni** [comment]
> Not in a rush with this, can wait a few days for sure. Can also update afterwards up to the v1.8.1 release.

v0.1.1 tag created: https://github.com/joinmarket-webui/jam/tree/v0.1.1

**Comment by openoms** [comment]
Rebased on `dev`, tested ok and no more changes are planned unless some finds 
something.
If you run a blitz go ahead and test this PR: 3321 or the `openoms` - `jamfix` branch.

Otherwise ready to be merged.

**Comment by openoms** [comment]
rebased to the latest `dev`

**Comment by openoms** [comment]
ok, the time has come. New fixes to a new PR please.

**Pull request merged**